### ISSUE TITLE
Allow different label sets for metric with the same name

### ISSUE THREAD

**Issue by lampad** [prometheus/client_ruby#318]
Currently the prometheus client does not allow you to register multiple metrics with the same name but different labels:

```
irb(main):002> m1 = Prometheus::Client::Counter.new(:example, docstring: "Example one with one label", labels: { custom_label_one: "foo" })
=>
#<Prometheus::Client::Counter:0x0000ffff9d5ae108
...
irb(main):003> m2 = Prometheus::Client::Counter.new(:example, docstring: "Example with two labels", labels: { custom_label_two: "bar", custom_label_one: "baz" })
=>
#<Prometheus::Client::Counter:0x0000ffff97cd4488
...
irb(main):004> Prometheus::Client.registry.register(m1)
=>
#<Prometheus::Client::Counter:0x0000ffff9d5ae108
 @all_labels_preset=false,
 @docstring="Example one with one label",
 @labels={:custom_label_one=>"foo"},
 @name=:example,
 @preset_labels={},
 @store=#<Prometheus::Client::DataStores::Synchronized::MetricStore:0x0000ffff973f5e48 @internal_store={}, @lock=#<Monitor:0x0000ffff973f5d58>>,
 @store_settings={},
 @validator=#<Prometheus::Client::LabelSetValidator:0x0000ffff973f6578 @expected_labels=[[:custom_label_one, "foo"]], @reserved_labels=[:pid]>>
irb(main):005> Prometheus::Client.registry.register(m2)
(irb):5:in `<main>': example has already been registered (Prometheus::Client::Registry::AlreadyRegisteredError)
```

[This has been discussed elsewhere](https://github.com/prometheus/client_java/issues/696) without any real conclusion, but after looking through the code, it seems relatively simple to allow this functionality in the ruby client registry by building a natural key of the metric name and possible label keys before sticking it in the registry.

Allowing this flexibility in label setting would make monitoring apps in distributed settings much easier and allow for more flexibility from a user/client standpoint by not having to know all labels up front.

I'd be happy to submit a PR if there's appetite for this change.

**Comment by Sinjo** [comment]
Hey @lampad, thanks for writing this up!

I'm going to take some time to read through the thread you linked and the related Prometheus docs. My gut take from a quick look is that this isn't something we're going to unilaterally decide over here in `client_ruby`. It's a convention agreed by the Prometheus team and implemented equally in the different client libraries.

If that ends up being the conclusion I come to after more reading, and you would like to pursue the wider change, I'd suggest raising this for a discussion on the [prometheus-developers mailing list](https://groups.google.com/g/prometheus-developers).

**Comment by lampad** [comment]
I think this will probably be the piece we'll end up discussing the most [from the docs](https://prometheus.io/docs/instrumenting/writing_clientlibs/#labels):

> Client libraries MUST NOT allow users to have different label names for the same metric for Gauge/Counter/Summary/Histogram or any other Collector offered by the library.

One could argue semantics of a metric with the same name but different labels being a different metric altogether, but I think I'd also like to challenge that general concept of disallowing different label names on a metric, or at least get a bit more context as to why this seems to be such a firm directive.

To me it seems really useful to allow for varying labels as a method of grouping metrics in different ways for reporting and observability. In a distributed setting, for different components producing the same thing in different contexts it seems useful that those components wouldn't need to know the labels being set by their counterpart if they're incrementing a counter with the same name for example.

In any case, interested to hear your thoughts after you do a bit more reading.

**Comment by lampad** [comment]
> So this bit I'm slightly confused by.

Yeah that's totally fair, and I realized that it is per-process after posting it so this was me just being dumb. I do think it's interesting that multiple same-named metrics with different labels are allowed at the collector level, so it's a bit confusing as to why it wouldn't be allowed intra-process.

I'll raise it to the group (obviously after organizing my thoughts a bit more lol). Appreciate the feedback.
